clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslServerAuthenticator.java

```diff
@@ -301,6 +327,53 @@ public void close() throws IOException {
             saslServer.dispose();
     }
 
+    @Override
+    public void reauthenticate(ReauthenticationContext reauthenticationContext) throws IOException {
+        NetworkReceive saslHandshakeReceive = reauthenticationContext.networkReceive();
+        if (saslHandshakeReceive == null)
+            throw new IllegalArgumentException(
+                    "Invalid saslHandshakeReceive in server-side re-authentication context: null");
+        Authenticator previousAuthenticator = reauthenticationContext.previousAuthenticator();
+        if (!(previousAuthenticator instanceof SaslServerAuthenticator))
+            throw new IllegalArgumentException(
+                    "Invalid previous authenticator in server-side re-authentication context: "
+                            + previousAuthenticator);
+        SaslServerAuthenticator previousSaslServerAuthenticator = (SaslServerAuthenticator) previousAuthenticator;
+        previousSaslMechanism = previousSaslServerAuthenticator.saslMechanism();
+        previousKafkaPrincipal = previousSaslServerAuthenticator.principal();
+        previousSaslServerAuthenticator.close();
+        netInBuffer = saslHandshakeReceive;
+        reauthenticating = true;
+        reauthenticationBeginNanos = reauthenticationContext.reauthenticationBeginNanos();
+        LOG.debug("Beginning re-authentication: {}", this);
+        netInBuffer.payload().rewind();
+        processPayload();
```

Comment:
We should be able to set up the state and call `authenticate` without separating out the methods.

Reply:
👍 I added a new `REAUTH_PROCESS_HANDSHAKE` state and that is where re-authentication starts.